clang/include/clang/CIR/Dialect/IR/CIROps.td

```diff
@@ -2326,6 +2326,22 @@ def ThrowOp : CIR_Op<"throw",
   let hasVerifier = 1;
 }
 
+def StackSaveOp : CIR_Op<"stack_save"> {
+  let summary = "remember the current state of the function stack";
+  let description = [{ remember the current state of the function stack }];
+
+  let results = (outs AnyType:$result);
+  let assemblyFormat = "attr-dict `:` type($result)";
+}
+
+def StackRestoreOp : CIR_Op<"stack_restore"> {
+  let summary = "restore the state of the function stack";
+  let description = [{ restore the state of the function stack }];
+
+  let arguments = (ins CIR_PointerType:$ptr);
+  let assemblyFormat = "$ptr attr-dict `:` qualified(type($ptr))";
+}
+
```

Comment:
If I understand you right, we need to emit stack save/restore during the lowering to LLVM IR dialect. But we would need to analyze nearly the whole function body in this case: any nested block with return need to be updated manually with `stack_restore` operation, as far I understand. And to infer what the right value we need to pass into it.  For me it sounds like we need to perform some analysis in this case, but not just replace one to another.

Reply:
> problem I keep fighting with: emit cleanup exactly where it need to be done. But I still don't see how your suggestion will help us 

Keep in mind that my suggestion is to avoid redundancy and code bloat. It will not necessarily facilitate identifying where the setup (`stackstore`) and clean up (`stackrestore`) should be emitted.

> probably I don'y understand it clearly.

The core of the idea is to **avoid redundancy/code-bloat in CIR by using its high-level information** (in this case, scopes). We already know where to save and restore the stack by evaluating the scope.  We could manually add these operations to it, but they would be redundant and complicate the IR.

> if I understand you right, we need to emit stack save/restore during the lowering to LLVM IR dialect.

So far so good!

> we would need to analyze nearly the whole function body in this case

This part I disagree. We would work with two insertion points for each scope that has a dynamic allocation: after static allocations in a scope, we emit the `stacksave`, and right before the end of the scope, we emit the `stackrestore`.

Reply:
The next thing makes me doubt a little. But the idea looks very interesting. 
Look, let's take a dummy code as an example. I didn't post CIR code for the brevity :) 
```
int foo(int len, int c) {
    int a[len];  // VLA
    init(len, a);     
    if (c > 5) {
        if (c < 10) {
            do_some_1(len, a);
            return 0; // stack restore needed before
        } else {
            do_some_2(len, a);
            return 1;  // stack restore needed before
        }
    }
    else if (c < 3) {
        do_some_3(len, a);
        // yield from the scope, but no return - no stack restore here
    }

    return 42; // stack restore needed before
}
```
> We would work with two insertion points for each scope that has a dynamic allocation

Looks like there are more then two insertion points.
Each `return` need to be preceded by the `stack_restore`, right?   How it would look like in CIR?
First, save the stack
```
    %1 = cir.alloca !cir.ptr<!u8i>, cir.ptr <!cir.ptr<!u8i>>, ["saved_stack"] {alignment = 8 : i64} 
    ... 
    %5 = cir.stack_save : <!u8i> 
    cir.store %5, %1 : !cir.ptr<!u8i>, cir.ptr <!cir.ptr<!u8i>> 
```
And then, restore it before each return.
```
    ....
    %41 = cir.load %1 : cir.ptr <!cir.ptr<!u8i>>, !cir.ptr<!u8i> 
    cir.stack_restore %41 : !cir.ptr<!u8i> 
```
Something similar we'll have to do during the lowering. So we'll have to walk through the scopes and precede each `return` with the operations above, i.e. `analyze`  nested scopes as well, check if there is a yielding from the current scope and insert `stack_restore`. Don't do anything for the yielding in the nested scopes, unless there a return from function or from loop body (i.e. break)...

This is a problem I'm talking about - but again, looks like it's doable.  

Next, we have to seek for the dynamic allocations in the regions in order to insert `stack_save` stuff  - not only in the `ScopeOp`, but in the function body, in the `loop` body, in the `if ` regions  - and we can easily miss where else. So we'll have to perform search in every operation that contains regions. This makes me think, that we would need to make changes in many operations in the lowering. May be there is an easier way and I just don't know it.

Reply:
> Looks like there are more then two insertion points. Each `return` need to be preceded by the `stack_restore`

Good point! Lemme see if we got all cases where `stackrestore` has to be called:

 - Restore before `cir.yield break` if the current scope has a dynamic allocation.
 - Restore before `cir.yield continue` if the current scope has a dynamic allocation.
 - Restore before `cir.return` if the **current or any parent scopes** have a dynamic allocation.
 
Any other cases that you can think of?

> So we'll have to perform search in every operation that contains regions

> May be there is an easier way and I just don't know it.
 
I believe there's a simpler way.  Instead of modifying the lowering of operations that define the scope (e.g. `cir.if`, `cir.loop`, etc.), let's focus on the ones that mark where the `stacksave`/`stackrestore` ought to be called (e.g. `cir.alloca`, `cir.yield`, `cir.return`).

>  we'll have to walk through the scopes and precede each `return` with the operations above

We could just modify the lowering of the `cir.yield` and `cir.return` operations to evaluate if a `stackrestore` has to be called by analyzing parent regions/operations. The first just requires us to load the result of the `stacksave` of the current scope, the latter is trickier since we need to look up scope-by-scope for the result of the **first** `stacksave` called in the function, but I believe both can easily be done with MLIR APIs without too much overhead.

> Next, we have to seek for the dynamic allocations in the regions in order to insert stack_save stuff - not only in the ScopeOp, but in the function body, in the loop body, in the if  regions

In this case, whenever a pattern rewriter for a `cir::AllocaOp` is called:

 - Evaluate if it is a dynamic or static allocation.
 - If it is dynamic, check if the `stacksave` has already been emitted in the current scope, if not, emit it.
 - Proceed to lower the dynamic allocation.

> So.. what do you think?

If we think about changing the lowering of operations that define the scope, it seems a bit daunting, but looking only at the `cir.alloca`, `cir.yield`, and `cir.return` operations, not so much.

Does that make sense?

Reply:
sounds reasonable!

Reply:
@sitio-couto 
What do you think, may be it would be better to perform all the above in the `LoweringPrepare` stage? in this case we would need these operations in cir though. 
The reason I think about it is that toooo much `YieldOp` in the `LowerToLLVM`, and we need to insert our checks  everywhere, I mean checks if the stack restore need to be inserted. And looks like we can't just add a separate lowering  for `YieldOp` since every this operation is deleted by the parent ops, like `LoopOp` or `IfOp`. Or .. there is a chance to preprocess yields first somehow ?

Reply:
> @gitoleg perhaps we should go with your initial implementation of just adding `cir.stacksave` and `cir.stackrestore` during codegen.

+1, let's start simple! We can find a way to embed this info as part of cir.scope later on, but I'd like to fix dtors in general first because overall they should be CIRGen'd in the same place.